### PR TITLE
ReplicationOperation should fail gracefully

### LINKED ISSUE #97183: [CI] IndexRecoveryIT testPeerRecoveryTrimsLocalTranslog failing
The failing assertion was added recently in #97133. I'll suppress it while investigating.

**Build scan:**
https://gradle-enterprise.elastic.co/s/qlsep6gfoto36/tests/:server:internalClusterTest/org.elasticsearch.indices.recovery.IndexRecoveryIT/testPeerRecoveryTrimsLocalTranslog

**Reproduction line:**
```
./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.indices.recovery.IndexRecoveryIT.testPeerRecoveryTrimsLocalTranslog" -Dtests.seed=6C5DEBB10362A47D -Dtests.locale=mt-MT -Dtests.timezone=America/Thule -Druntime.java=20
```

**Applicable branches:**
main

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.indices.recovery.IndexRecoveryIT&tests.test=testPeerRecoveryTrimsLocalTranslog

**Failure excerpt:**
```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=4364, name=elasticsearch[node_t1][write][T#2], state=RUNNABLE, group=TGRP-IndexRecoveryIT]

  at __randomizedtesting.SeedInfo.seed([6C5DEBB10362A47D:91DBA9F3E716E8A1]:0)

  Caused by: java.lang.AssertionError: (No message provided)

    at __randomizedtesting.SeedInfo.seed([6C5DEBB10362A47D]:0)
    at org.elasticsearch.index.shard.IndexShard.syncAfterWrite(IndexShard.java:3630)
    at org.elasticsearch.action.support.replication.TransportWriteAction$AsyncAfterWriteAction.run(TransportWriteAction.java:518)
    at org.elasticsearch.action.support.replication.TransportWriteAction$WritePrimaryResult.runPostReplicationActions(TransportWriteAction.java:318)
    at org.elasticsearch.action.support.replication.ReplicationOperation.handlePrimaryResult(ReplicationOperation.java:173)
    at org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:169)
    at org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:94)
    at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:319)
    at org.elasticsearch.action.bulk.TransportShardBulkAction$2.finishRequest(TransportShardBulkAction.java:273)
    at org.elasticsearch.action.bulk.TransportShardBulkAction$2.doRun(TransportShardBulkAction.java:235)
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.performOnPrimary(TransportShardBulkAction.java:286)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:137)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:74)
    at org.elasticsearch.action.support.replication.TransportWriteAction$1.doRun(TransportWriteAction.java:215)
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
    at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
    at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
    at java.lang.Thread.run(Thread.java:1623)

```

### PR DESCRIPTION
**Pull request by kingherc**
Problem:
finishAsFailed could be called before, or even asynchronously in the middle of, operations like runPostReplicationActions which try to sync the translog. finishAsFailed immediately triggers the failure of the resultListener which releases the index shard primary operation permit. This means that runPostReplicationActions may try to sync the translog without an operation permit.

Solution:
We refactor the infrastructure of ReplicationOperation regarding pendingActions and the resultListener, by replacing them with a RefCountingListener. This way, if there are failures, the result listener is called once, after all pending actions are done.

For the specific error we got in issue #97183, this means that a call to onNoLongerPrimary (which can happen if we fail to fail a replica shard or mark it as stale) will not immediately release the primary operation permit and the assertion in the translog sync will be honored.

Fixes #97183

### CODE REVIEW

**Comment by Tim-Brooks** [comment]
I should be able to dig into this tomorrow.

File: server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java [hunk 1]
```diff
@@ -106,28 +113,33 @@ public ReplicationOperation(
...
+                    );
+                return;
+            }
 
-        totalShards.incrementAndGet();
-        pendingActions.incrementAndGet(); // increase by 1 until we finish all primary coordination
-        primary.perform(request, ActionListener.wrap(this::handlePrimaryResult, this::finishAsFailed));
+            totalShards.incrementAndGet();
+            var primaryCoordinationPendingActionListener = pendingActionsListener.acquire(); // triggered when we finish all coordination
+            primary.perform(request, ActionListener.wrap(primaryResult -> {
```
Comment (Tim-Brooks) [inline server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:136]: Maybe

```java
            totalShards.incrementAndGet();
            var primaryCoordinationPendingActionListener = pendingActionsListener.acquire(); // triggered when we finish all coordination
            primary.perform(request, primaryCoordinationPendingActionListener.delegateFailureAndWrap(this::handlePrimaryResult));
        }
    }

    private void handlePrimaryResult(ActionListener<Void> primaryCoordinationPendingActionListener, final PrimaryResultT primaryResult) {
        this.primaryResult = primaryResult;
        final ReplicaRequest replicaRequest = primaryResult.replicaRequest();
        if (replicaRequest != null) {
            if (logger.isTraceEnabled()) {
                logger.trace("[{}] op [{}] completed on primary for request [{}]", primary.routingEntry().shardId(), opType, request);
            }
            final ReplicationGroup replicationGroup = primary.getReplicationGroup();
```

File: server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java [hunk 2]
```diff
@@ -193,20 +205,28 @@ public void onFailure(Exception e) {
...
-            pendingActions.incrementAndGet();
+            var staleCopyPendingActionListener = pendingActionsListener.acquire();
             replicasProxy.markShardCopyAsStaleIfNeeded(
                 replicaRequest.shardId(),
                 allocationId,
                 primaryTerm,
-                ActionListener.wrap(r -> decPendingAndFinishIfNeeded(), ReplicationOperation.this::onNoLongerPrimary)
+                ActionListener.wrap(
+                    r -> staleCopyPendingActionListener.onResponse(null),
+                    e -> onNoLongerPrimary(e, staleCopyPendingActionListener)
```
Comment (Tim-Brooks) [inline server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:228]: Maybe

```
staleCopyPendingActionListener.delegateResponse((listener, e) -> onNoLongerPrimary(e, listener))
```

File: server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java [hunk 3]
```diff
@@ -282,7 +307,10 @@ public void onFailure(Exception replicaException) {
                     primaryTerm,
                     message,
                     replicaException,
-                    ActionListener.wrap(r -> decPendingAndFinishIfNeeded(), ReplicationOperation.this::onNoLongerPrimary)
+                    ActionListener.wrap(
```
Comment (Tim-Brooks) [inline server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:310]: Maybe 

```java
replicationPendingActionListener.delegateResponse(
                        (listener, e) -> onNoLongerPrimary(e, replicationPendingActionListener)
                    )
```

File: server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java [hunk 4]
```diff
@@ -369,13 +397,13 @@ public void onAfter() {
         }
     }
 
-    private void onNoLongerPrimary(Exception failure) {
+    private void onNoLongerPrimary(Exception failure, ActionListener<Void> pendingActionlistener) {
```
Comment (Tim-Brooks) [inline server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:400]: Maybe just named titled `listener` to more differentiate between this being local scoped and the field name.

**Review by Tim-Brooks — APPROVED** [review]
LGTM. All the comments are optional. Just things around listener methods.

**Comment by kingherc** [comment]
Thanks @Tim-Brooks ! Incorporated your feedback, and also made some nit changes to ensure no exception escapes through (by making use of try-with-resources for the ActionListenerRef and some ActionListener.runs in a couple of places). Feel free to give another quick look if you'd like.

**Comment by Tim-Brooks** [comment]
> Feel free to give another quick look if you'd like.

LGTM

**Pull request merged**